store/src/java/com/zimbra/cs/service/admin/Auth.java

```diff
@@ -185,135 +185,139 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
-                                    boolean verifyAccount = authTokenEl.getAttributeBool(AccountConstants.A_VERIFY_ACCOUNT, false);
-                                    if (verifyAccount && !twoFactorTokenAcct.getId().equalsIgnoreCase(acct.getId())) {
-                                        throw new AuthTokenException("two-factor auth token doesn't match the named account");
-                                    }
-                                } catch (AuthTokenException e) {
-                                    AuthFailedServiceException exception = AuthFailedServiceException
-                                        .AUTH_FAILED("bad auth token");
-                                    AuthListener.invokeOnException(exception);
-                                    throw exception;
-                                }
-                            }
-                            TwoFactorAuth manager = TwoFactorAuth.getFactory().getTwoFactorAuth(acct);
-                            if (twoFactorCode != null) {
-                                manager.authenticate(twoFactorCode);
-                            } else {
-                                AuthFailedServiceException e = AuthFailedServiceException
-                                    .AUTH_FAILED("no two-factor code provided");
-                                AuthListener.invokeOnException(e);
-                                throw e;
-                            }
-                            if (twoFactorToken != null) {
-                                try {
-                                    twoFactorToken.deRegister();
-                                } catch (AuthTokenException e) {
-                                    throw ServiceException.FAILURE("cannot de-register two-factor auth token", e);
-                                }
-                            }
-                        }
-                    }
-                } 
-
-                if (registerTrustedDevice && (trustedToken == null || trustedToken.isExpired())) {
-                    //generate a new trusted device token if there is no existing one or if the current one is no longer valid
-                    Map<String, Object> attrs = getTrustedDeviceAttrs(zsc, newDeviceId == null? deviceId: newDeviceId);
-                    TrustedDevices trustedDeviceManager = TwoFactorAuth.getFactory().getTrustedDevices(acct);
-                    if (trustedDeviceManager != null) {
-                        trustedToken = trustedDeviceManager.registerTrustedDevice(attrs);
-                    }
-                }
+                // Only perform 2fa authentication on admin account not on the resource accounts
+                if(acct != null && !Boolean.parseBoolean(acct.getAttr(AccountConstants.E_IS_SYSTEM_RESOURCE))) {
+	                String deviceId = request.getAttribute(AccountConstants.E_DEVICE_ID, null);
```

Comment:
I prefer to put entire block into private method

Reply:
doing this